### PR TITLE
Introduces new API for accessing price data:

### PR DESCRIPTION
**Pull request by jbredeche**
data.get_spot_value
data.can_trade
data.is_stale
data.history

### CODE REVIEW

**Comment by jbredeche** [comment]
The `test_history.py` diff is huge because I completely rewrote the history tests.

File: tests/test_bar_data.py [hunk 1]
```diff
@@ -0,0 +1,418 @@
+from unittest import TestCase
```
Comment (ehebert) [inline tests/test_bar_data.py]: Copyright.

File: tests/test_bar_data.py [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+from zipline.utils.test_utils import write_minute_data_for_asset, \
+    create_daily_df_for_asset, DailyBarWriterFromDataFrames, assert_same
+
+OHLC = ["open", "high", "low", "close"]
+OHLCP = OHLC + ["price"]
+ALL_FIELDS = OHLCP + ["volume", "last_traded"]
+
+# offsets used in test data
+field_info = {
+    "open": 1,
```
Comment (ehebert) [inline tests/test_bar_data.py]: Could we make these mantissas instead?
So that same value is not in different slots for OHLC. e.g. low of i, and close of i+1 would have the same value.

File: zipline/_protocol.pyx [hunk 3]
```diff
@@ -76,6 +79,289 @@ cdef class BarData:
...
+                    self.simulation_dt_func(),
+                    self.data_frequency
+                )
+            else:
+                # assume fields is iterable
+                # return a Series indexed by field
+                return pd.Series(data={
+                    field: self.data_portal.get_spot_value(
+                        assets,
+                        field,
```
Comment (ehebert) [inline zipline/_protocol.pyx]: We should make (but not required for this PR) `get_spot_value` accept the iterable as well.

File: zipline/_protocol.pyx [hunk 4]
```diff
@@ -76,6 +79,289 @@ cdef class BarData:
...
+                        self.simulation_dt_func(),
+                        self.data_frequency)
+                    for field in fields
+                }, index=fields, name=assets.symbol)
+        else:
+            if isinstance(fields, str):
+                # assume assets is iterable
+                # return a Series indexed by asset
+                return pd.Series(data={
+                    asset: self.data_portal.get_spot_value(
```
Comment (ehebert) [inline zipline/_protocol.pyx]: If `get_spot_value` returns the values in the same order as they are passed in, the returned array from `get_spot_value` can be passed to the Series without the extra step of the dict.

File: zipline/_protocol.pyx [hunk 5]
```diff
@@ -76,6 +79,289 @@ cdef class BarData:
...
+                        assets,
+                        field,
+                        self.simulation_dt_func(),
+                        self.data_frequency)
+                    for field in fields
+                }, index=fields, name=assets.symbol)
+        else:
+            if isinstance(fields, str):
+                # assume assets is iterable
+                # return a Series indexed by asset
```
Comment (ehebert) [inline zipline/_protocol.pyx]: What do you think of doing a `field = fields` here, just so that scalar/iterable is clearer.

File: zipline/_protocol.pyx [hunk 6]
```diff
@@ -76,6 +79,289 @@ cdef class BarData:
...
+                        self.simulation_dt_func(),
+                        self.data_frequency)
+                    for asset in assets
+                    }, index=assets, name=fields)
+            else:
+                # both assets and fields are iterable
+                data = {}
+
+                for field in fields:
+                    series = pd.Series(data={
```
Comment (ehebert) [inline zipline/_protocol.pyx]: This works, but I think having `get_spot_value` return a ndarray may help here.

File: zipline/_protocol.pyx [hunk 7]
```diff
@@ -76,6 +79,289 @@ cdef class BarData:
...
+            })
+
+    cdef _is_stale_for_asset(self, asset, dt, data_portal):
+        if asset.start_date > dt:
+            return False
+
+        if asset.end_date <= dt:
+            return False
+
+        current_val = data_portal.get_spot_value(asset, "close", dt,
```
Comment (ehebert) [inline zipline/_protocol.pyx]: I'd recommend using `volume` since a check against 0 is cheaper.

File: zipline/data/data_portal.py [hunk 8]
```diff
@@ -286,41 +286,45 @@ def get_spot_value(self, asset, field, dt, data_frequency):
...
-            field,
-            dt,
-        )
-
-        if extra_source_val is not None:
-            return extra_source_val
-
-        if field not in BASE_FIELDS:
-            raise KeyError("Invalid column: " + str(field))
+        # extra_source_val = self._check_extra_sources(
```
Comment (ehebert) [inline zipline/data/data_portal.py]: Remove?

File: zipline/data/data_portal.py [hunk 9]
```diff
@@ -35,16 +35,17 @@
...
-    'open_price': 'open',
-    'high': 'high',
-    'low': 'low',
-    'close': 'close',
-    'close_price': 'close',
-    'volume': 'volume',
-    'price': 'close'
-}
+BASE_FIELDS = frozenset([
+    "open", "high", "low", "close", "volume", "price", "last_traded"
```
Comment (ehebert) [inline zipline/data/data_portal.py]: Is `open_price` intentionally no longer supported with this change?

File: tests/test_algorithm.py [hunk 10]
```diff
@@ -706,7 +706,7 @@ def setUpClass(cls):
         for sid in cls.sids:
             equities_metadata[sid] = {
                 'start_date': cls.sim_params.period_start,
-                'end_date': cls.sim_params.period_end
+                'end_date': cls.env.next_trading_day(cls.sim_params.period_end)
```
Comment (ehebert) [inline tests/test_algorithm.py]: Why does the equity need to end beyond the trading range?

File: tests/test_bar_data.py [hunk 11]
```diff
@@ -0,0 +1,429 @@
...
+        self.assertEqual(4, bar_data.spot_value(self.ASSET1, "high"))
+        self.assertEqual(1, bar_data.spot_value(self.ASSET1, "low"))
+        self.assertEqual(2, bar_data.spot_value(self.ASSET1, "close"))
+        self.assertEqual(200, bar_data.spot_value(self.ASSET1, "volume"))
+        self.assertEqual(2, bar_data.spot_value(self.ASSET1, "price"))
+        self.assertEqual(self.days[0],
+                         bar_data.spot_value(self.ASSET1, "last_traded"))
+
+        for field in OHLCP:
+            self.assertTrue(np.isnan(bar_data.spot_value(self.ASSET2, field)))
```
Comment (ehebert) [inline tests/test_bar_data.py]: So that it can be seen which if any values fail. It would be helpful here to either include the field as the 3rd argument to the assert, or `np.testing.assert_array_equal` does support checking `np.nan`s against eachother, so this could be.

```
np.testing.assert_array_equal(
   [np.nan] * 5,
   [bar_data.spot_value(self.ASSET2, field) for field in OHLCP]
)
```

 

```

File: tests/test_blotter.py [hunk 12]
```diff
@@ -58,7 +58,9 @@ def setUpClass(cls):
         cls.env.write_data(equities_data={
             24: {
                 'start_date': cls.sim_params.trading_days[0],
-                'end_date': cls.sim_params.trading_days[-1]
+                'end_date': cls.env.next_trading_day(
```
Comment (ehebert) [inline tests/test_blotter.py]: Why is this the only change in this file?

File: zipline/_protocol.pxd [hunk 13]
```diff
@@ -18,6 +18,10 @@ cdef class BarData:
     cdef object simulation_dt_func
     cdef object data_frequency
     cdef dict _views
+    cdef object _get_equity_price_view(BarData, object)
+    cdef object _create_sid_view(BarData, object)
+    cdef object _is_stale_for_asset(BarData, object, object, object)
```
Comment (ehebert) [inline zipline/_protocol.pxd]: Could type these as `bool`.

(`from cpython import `bool`)

File: zipline/_protocol.pyx [hunk 14]
```diff
@@ -76,6 +79,289 @@ cdef class BarData:
...
+        -------
+        boolean or Series of booleans, indexed by asset.
+        """
+        dt = self.simulation_dt_func()
+        data_portal = self.data_portal
+
+        if isinstance(assets, Asset):
+            return self._can_trade_for_asset(assets, dt, data_portal)
+        else:
+            return pd.Series(data={
```
Comment (ehebert) [inline zipline/_protocol.pyx]: Have we done some profiling of whether a dict versus Series would be more performant here?

File: zipline/_protocol.pyx [hunk 15]
```diff
@@ -76,6 +79,289 @@ cdef class BarData:
...
+                asset: self._can_trade_for_asset(asset, dt, data_portal)
+                for asset in assets
+            })
+
+    cdef _can_trade_for_asset(self, asset, dt, data_portal):
+        if asset.start_date > dt:
```
Comment (ehebert) [inline zipline/_protocol.pyx]: I'd recommend evaluating to the positive first.

```
if asset.start_date <= dt <= self.asset.end_date:
    return data_portal.get_spot_value(asset, 'volume', dt, self._data_frequency) != 0
else:
    return False
```

File: zipline/_protocol.pyx [hunk 16]
```diff
@@ -76,6 +79,289 @@ cdef class BarData:
...
+            return False
+
+        if asset.end_date < dt:
+            return False
+
+        # is there a last price?
+        return not np.isnan(
+            data_portal.get_spot_value(asset, "price", dt, self.data_frequency)
```
Comment (ehebert) [inline zipline/_protocol.pyx]: Would recommend checking volume here since a comp against 0 is cheaper than `nan`.

File: zipline/_protocol.pyx [hunk 17]
```diff
@@ -76,6 +79,289 @@ cdef class BarData:
...
+        if isinstance(assets, Asset):
+            return self._is_stale_for_asset(assets, dt, data_portal)
+        else:
+            return pd.Series(data={
+                asset: self._is_stale_for_asset(asset, dt, data_portal)
+                for asset in assets
+            })
+
+    cdef _is_stale_for_asset(self, asset, dt, data_portal):
+        if asset.start_date > dt:
```
Comment (ehebert) [inline zipline/_protocol.pyx]: Can you put a docstring here about the semantics of `stale`? `stale` means no trade for  data for the current bar, correct?

Do we need the `asset.start_date` check here? Returning `False` when there is no data seems counter intuitive, and this function is only needed if and only if the stock is 'active' correct?

That is, we are incurring extra dt comps that may be unneeded.

File: zipline/_protocol.pyx [hunk 18]
```diff
@@ -76,6 +79,289 @@ cdef class BarData:
...
+                asset_list,
+                self.simulation_dt_func(),
+                bar_count,
+                frequency,
+                fields
+            )
+
+            if single_asset:
+                # single asset, single field, return a series.
+                return df[assets]
```
Comment (ehebert) [inline zipline/_protocol.pyx]: Possible TODO: Make `get_history_window` return just the one asset if called with only one value.

File: zipline/_protocol.pyx [hunk 19]
```diff
@@ -76,6 +79,289 @@ cdef class BarData:
...
+                    field: self.data_portal.get_history_window(
+                        [assets],
+                        self.simulation_dt_func(),
+                        bar_count,
+                        frequency,
+                        field
+                    )[assets] for field in fields
+                })
+            else:
+                df_dict = {
```
Comment (ehebert) [inline zipline/_protocol.pyx]: Have not profiled, but I the extra DataFrame creation per sid at a large number will be expensive. We should look at either having `get_history_window` return a `Panel` or 3-dim ndarray.

File: zipline/algorithm.py [hunk 20]
```diff
@@ -786,8 +786,18 @@ def _calculate_order_value_amount(self, asset, value):
...
+        if self.datetime >= asset.end_date:
+            raise CannotOrderDelistedAsset(
+                msg="Cannot order {0}, as it stopped trading on {1}.".format(
+                    asset.symbol, asset.end_date
+                )
+            )
         last_price = self.trading_client.current_data[asset].price
 
+        if np.isnan(last_price):
+            raise
```
Comment (ehebert) [inline zipline/algorithm.py]: Should name an exception here.

File: zipline/data/data_portal.py [hunk 21]
```diff
@@ -429,24 +429,25 @@ def _get_minute_spot_value_future(self, asset, column, dt):
...
-            asset.sid,
-            dt,
-            column
+            asset.sid, dt, column
         )
 
-        if result > 0 or column == "volume":
-            return result
+        if result == 0 or np.isnan(result):
+            if column == "volume":
```
Comment (ehebert) [inline zipline/data/data_portal.py]: I'd recommend restructuring so that volume and OHLC handling are not comingled.
With the implementation above when `volume` is non-zero `np.isnan(result)` is also called.

That extra call would be removed with:

```
if column == "volume":
     if result == 0:
        return 0
elif np.isnan(result) and not ffill:
       return  np.nan
```

File: zipline/data/data_portal.py [hunk 22]
```diff
@@ -717,27 +742,28 @@ def get_history_window(self, assets, end_dt, bar_count, frequency, field,
...
-        assets = np.array([
-            (self.env.asset_finder.retrieve_asset(asset) if
-             isinstance(asset, int) else asset) for asset in assets])
+        if field not in OHLCVP_FIELDS:
+            raise ValueError("Invalid field: {0}").format(field)
 
         if frequency == "1d":
-            df = self._get_history_daily_window(assets, end_dt, bar_count,
-                                                field_to_use)
+            if field == "price":
```
Comment (ehebert) [inline zipline/data/data_portal.py]: To remove some duplication at the beginning of the function could do:
```
if field == 'price':
    field = 'close'
```

File: zipline/data/data_portal.py [hunk 23]
```diff
@@ -1017,7 +1050,8 @@ def _get_daily_window_for_sid(self, asset, field, days_in_window,
         else:
             return_array = np.zeros((bar_count,))
 
-        return_array[:] = np.NAN
+        if field != "volume":
+            return_array[:] = np.NAN
```
Comment (ehebert) [inline zipline/data/data_portal.py]: We do nothing with volume because the array is already zeros, correct? Maybe a comment to that effect?

File: zipline/data/us_equity_pricing.py [hunk 24]
```diff
@@ -528,6 +528,16 @@ def _spot_col(self, colname):
     def get_last_traded_dt(self, asset, day):
         volumes = self._spot_col('volume')
         search_day = day
+
+        if search_day >= asset.end_date:
+            # move to the last day before the asset end date
+            try:
+                search_day = self._calendar[
+                    self._calendar.get_loc(asset.end_date) - 1
```
Comment (ehebert) [inline zipline/data/us_equity_pricing.py]: I believe this can also be accomplished by `asset.end_date - self._calendar.freq`.
(Though not 100% if this calendar object has a `freq`).

File: zipline/data/us_equity_pricing.py [hunk 25]
```diff
@@ -528,6 +528,16 @@ def _spot_col(self, colname):
     def get_last_traded_dt(self, asset, day):
         volumes = self._spot_col('volume')
         search_day = day
+
+        if search_day >= asset.end_date:
+            # move to the last day before the asset end date
+            try:
```
Comment (ehebert) [inline zipline/data/us_equity_pricing.py]: Would `searchsorted(asset.end_date, 'left')` encapsulate this entire logic?

File: zipline/sources/benchmark_source.py [hunk 26]
```diff
@@ -35,16 +35,38 @@ def __init__(self, benchmark_sid, env, trading_days, data_portal,
...
-                self.data_portal
-            )
+            self._precalculated_series = \
+                self._initialize_precalculated_series(
+                    self.benchmark_asset,
+                    self.env,
+                    self.trading_days,
+                    self.data_portal
+                )
+        else:
```
Comment (ehebert) [inline zipline/sources/benchmark_source.py]: Why was this change needed with the new API?

File: zipline/utils/test_utils.py [hunk 27]
```diff
@@ -678,6 +678,58 @@ def write_bcolz_minute_data(env, days, path, df_dict):
...
 
+def write_minute_data_for_asset(env, writer, start_dt, end_dt, sid,
+                                interval=1, start_val=1):
+
+    asset_minutes = env.minutes_for_days_in_range(start_dt, end_dt)
+    minutes_count = len(asset_minutes)
+    minutes_arr = np.array(range(start_val, start_val + minutes_count))
+
+    df = pd.DataFrame({
+        "open": minutes_arr + 1,
```
Comment (ehebert) [inline zipline/utils/test_utils.py]: Same comment earlier about using `0.1, 0.2` instead.

File: zipline/utils/test_utils.py [hunk 28]
```diff
@@ -678,6 +678,58 @@ def write_bcolz_minute_data(env, days, path, df_dict):
...
+        "high": days_arr + 2,
+        "low": days_arr - 1,
+        "close": days_arr,
+        "volume": days_arr * 100,
+        "day": [day.value for day in days]
+    })
+
+    if interval > 1:
+        # only keep every 'interval' rows
+        for idx, minute in enumerate(days_arr):
```
Comment (ehebert) [inline zipline/utils/test_utils.py]: I think this should be `idx, daily`.
Also, can do this all at once like this:

```
q, r = divmod(len(days_arr), interval)
where_zero = np.tile([False] + [True] * (interval - 1), q + (1 if r else 0))
where_zero = where_zero[:len(days_arr)]
df[where] = 0
```

File: zipline/utils/test_utils.py [hunk 29]
```diff
@@ -678,6 +678,58 @@ def write_bcolz_minute_data(env, days, path, df_dict):
...
+    df = pd.DataFrame({
+        "open": minutes_arr + 1,
+        "high": minutes_arr + 2,
+        "low": minutes_arr - 1,
+        "close": minutes_arr,
+        "volume": 100 * minutes_arr,
+        "dt": asset_minutes
+    }).set_index("dt")
+
+    if interval > 1:
```
Comment (ehebert) [inline zipline/utils/test_utils.py]: See below for a technique to do the indexing only once.

File: zipline/_protocol.pyx [hunk 30]
```diff
@@ -76,6 +79,289 @@ cdef class BarData:
...
+        fields: string or iterable of string.  Valid values are "open", "high",
+            "low", "close", "volume", "price", and "last_traded".
+
+        bar_count: integer number of bars of trade data
+
+        frequency: string. "1m" for minutely data or "1d" for daily date
+
+        Returns
+        -------
+        series or DataFrame or Panel, depending on the dimensionality of
```
Comment (ehebert) [inline zipline/_protocol.pyx]: `Series`

File: zipline/_protocol.pyx [hunk 31]
```diff
@@ -76,6 +79,289 @@ cdef class BarData:
...
+
+            If multiple assets and single field are passed in, the returned
+            DataFrame is indexed by dt, and has assets as columns.
+
+            If a single asset and multiple fields are passed in, the returned
+            DataFrame is indexed by dt, and has fields as columns.
+
+            If multiple assets and multiple fields are passed in, the returned
+            Panel is indexed by field, has dt as the major axis, and assets
+            as the minor axis.
```
Comment (ehebert) [inline zipline/_protocol.pyx]: Does this table summarize this correctly?


|                  | scalar Asset                         | iterable Assets                                         |
|----------------|---------------------------------------|--------------------------------------------------------|
| scalar field | scalar (float or int)                   | `DataFrame(index=dts, columns=assets)                      |
| iterable fields | `DataFrame(index=dts, columns=fields)` | `Panel(index=fields, major_axis=dt, minor_axis=assets) `    |

File: zipline/data/us_equity_pricing.py (hunk 24 above)
Comment (ssanderson) [inline zipline/data/us_equity_pricing.py]: Do we want the last entry if `get_loc` raises a KeyError (which would happen if the search day isn't in the calendar)?  Wondering if the try/except block is too broad.

File: zipline/finance/blotter.py [hunk 32]
```diff
@@ -227,10 +232,9 @@ def get_transactions(self, bar_data):
         closed_orders = []
         transactions = []
 
-        for asset, asset_orders in iteritems(self.open_orders):
-            trade_bar = bar_data[asset]
-
-            for order, txn in self.slippage_func(trade_bar, asset_orders):
+        for sid, asset_orders in iteritems(self.open_orders):
+            asset = self.asset_finder.retrieve_asset(sid)
```
Comment (ssanderson) [inline zipline/finance/blotter.py]: This is making a SQL Query per order.  You probably want `retrieve_all` to do this in one query.

File: zipline/_protocol.pyx (hunk 31 above)
Comment (ehebert) [inline zipline/_protocol.pyx]: @ssanderson do you have an opinion on the appropriate values for index and columns, and what would be more natural usage in an algorithm?

I think we should change the order to:

|                  | scalar Asset                         | iterable Assets                                         |
|----------------|---------------------------------------|--------------------------------------------------------|
| scalar field | scalar (float or int)                   | `DataFrame(index=assets, columns=dts)                      |
| iterable fields | `DataFrame(index=fields, columns=dts)` | `Panel(index=assets, major_axis=fields, minor_axis=dts) `    |

For a usage pattern of:
```
window = data.history(symbol['FOO'], ['open', 'close'], 90, '1d')
a = window['close']

window = data.history([symbol['FOO', 'BAR'], 'close', 90, '1d')
b = window['FOO']

window = data.history([symbol['FOO', 'BAR'], ['open', 'close'], 90, '1d')
c = window['FOO', 'close']

a == b == c
```

I can be convinced that `window['close', 'FOO']` is a more intuitive ordering, but recommended doing `[asset, field]` to preserver the ordering of the parameters to the history method.

File: zipline/finance/slippage.py [hunk 33]
```diff
@@ -49,27 +49,29 @@ def volume_for_bar(self):
         return self._volume_for_bar
 
     @abc.abstractproperty
-    def process_order(self, trade_bar, order):
+    def process_order(self, bar_data, price, volume, order):
```
Comment (ssanderson) [inline zipline/finance/slippage.py]: Unrelated to this PR, but it looks like this whole file is indented one space?  That's not the case on master.

File: zipline/_protocol.pyx (hunk 31 above)
Comment (ssanderson) [inline zipline/_protocol.pyx]: @ehebert I would transpose most of your suggestions.

The almost-universal convention for DataFrames is to use dt as the index, because the expectation is that you're generally accessing time-series values sequentially, and pandas uses columnar storage.  The reprs of dates are also much longer than most column names, which makes them inconvenient to display as column labels.  This is what I'd expect for the labels here:

|                  | scalar Asset                         | iterable Assets                                         |
|----------------|---------------------------------------|--------------------------------------------------------|
| scalar field | scalar (float or int)                   | `DataFrame(index=dts, columns=assets)                      |
| iterable fields | `DataFrame(columns=fields, index=dts)` | `Panel(items=fields, major_axis=dts, minor_axis=assets) `    |

The reason you want `fields` as the first index of a panel is that you guarantee that slices along the first axis yield blocks with homogenous storage. (All entries for a given field are guaranteed to have the same dtype.)  You then want `dt, asset` as major and minor axes so that slices by field yield `DataFrame`s with dates as the index.

Comment (ehebert) [inline zipline/_protocol.pyx]: @ssanderson you've convinced, we should be mindful of how the contiguous and have that take priority on the API.

File: zipline/_protocol.pyx (hunk 4 above)
Comment (jbredeche) [inline zipline/_protocol.pyx]: `data_portal.get_spot_value` doesn't return an array, it returns scalar

File: zipline/_protocol.pyx (hunk 5 above)
Comment (jbredeche) [inline zipline/_protocol.pyx]: good idea, done

File: zipline/_protocol.pyx (hunk 6 above)
Comment (jbredeche) [inline zipline/_protocol.pyx]: I'm open to it, but not optimizing for the multi-asset, multi-field case yet.  can do that later.

File: zipline/data/data_portal.py (hunk 9 above)
Comment (jbredeche) [inline zipline/data/data_portal.py]: yeah.  the new API uses "open" and "close".

BarData will do the translation from `open_price` to `open` and `close_price` to `close` to support the old API.

File: tests/test_algorithm.py (hunk 10 above)
Comment (jbredeche) [inline tests/test_algorithm.py]: because the mock data is written up to, and including, the last day of the trading range.  the asset's end date is the first day for which we don't have data.

the change to `price` to return `NaN` after the asset end date exposed this mismatch.

File: tests/test_bar_data.py (hunk 11 above)
Comment (jbredeche) [inline tests/test_bar_data.py]: good idea, done

File: tests/test_blotter.py (hunk 12 above)
Comment (jbredeche) [inline tests/test_blotter.py]: the price/NaN issue.  we need the asset to live to the day after the last trading day, so that we can get prices on the last trading day.

File: zipline/_protocol.pxd (hunk 13 above)
Comment (jbredeche) [inline zipline/_protocol.pxd]: thanks, done

File: zipline/_protocol.pyx (hunk 14 above)
Comment (jbredeche) [inline zipline/_protocol.pyx]: no.  can do it later.

File: zipline/_protocol.pyx (hunk 15 above)
Comment (jbredeche) [inline zipline/_protocol.pyx]: done, but your logic there isn't right (we need to check if there is a previous price, not if volume != 0)

File: zipline/_protocol.pyx (hunk 16 above)
Comment (jbredeche) [inline zipline/_protocol.pyx]: see note above - checking volume wouldn't be enough.

File: zipline/_protocol.pyx (hunk 17 above)
Comment (jbredeche) [inline zipline/_protocol.pyx]: the docstring is up in the main user-facing method, `is_stale`.

File: zipline/_protocol.pyx (hunk 19 above)
Comment (jbredeche) [inline zipline/_protocol.pyx]: agreed, can do later.

File: zipline/algorithm.py (hunk 20 above)
Comment (jbredeche) [inline zipline/algorithm.py]: oops, good catch.

File: zipline/data/data_portal.py (hunk 21 above)
Comment (jbredeche) [inline zipline/data/data_portal.py]: done

File: zipline/data/data_portal.py (hunk 22 above)
Comment (jbredeche) [inline zipline/data/data_portal.py]: that would break the logic later in the method that decides whether to ffill

File: zipline/data/data_portal.py (hunk 23 above)
Comment (jbredeche) [inline zipline/data/data_portal.py]: yes.  will comment.

File: zipline/finance/slippage.py [hunk 34]
```diff
@@ -145,8 +146,8 @@ def process_order(self, trade_bar, order):
                 return
 
         return create_transaction(
-            trade_bar,
             order,
+            bar_data.current_dt,
```
Comment (ssanderson) [inline zipline/finance/slippage.py]: It looks like this is the only place where this needs the `bar_data` object now.  Could we just pass the dt so that this function just operates on plain old data?

File: zipline/finance/slippage.py [hunk 35]
```diff
@@ -181,11 +182,11 @@ def __init__(self, spread=0.0):
         """
         self.spread = spread
 
-    def process_order(self, trade_bar, order):
+    def process_order(self, bar_data, price, volume, order):
         return create_transaction(
-            trade_bar,
             order,
-            trade_bar.price + (self.spread / 2.0 * order.direction),
+            bar_data.current_dt,
```
Comment (ssanderson) [inline zipline/finance/slippage.py]: Same note as above.

File: zipline/utils/test_utils.py [hunk 36]
```diff
@@ -678,6 +678,58 @@ def write_bcolz_minute_data(env, days, path, df_dict):
...
+    days_count = len(days)
+    days_arr = np.array(range(2, days_count + 2))
+
+    df = pd.DataFrame({
+        "open": days_arr + 1,
+        "high": days_arr + 2,
+        "low": days_arr - 1,
+        "close": days_arr,
+        "volume": days_arr * 100,
+        "day": [day.value for day in days]
```
Comment (ssanderson) [inline zipline/utils/test_utils.py]: I think this just wants to be `days.values`?

File: zipline/_protocol.pyx [hunk 37]
```diff
@@ -76,6 +81,289 @@ cdef class BarData:
...
+            "last_traded", "open", "high", "low", "close", "volume"
+
+        Returns
+        -------
+        Float, pandas Series, or pandas DataFrame.  See notes
+        below.
+
+        Notes
+        -----
+        If a single asset and a single field are passed in, a scalar float
```
Comment (ssanderson) [inline zipline/_protocol.pyx]: This might be a case where putting a table in here like the one @ehebert commented with would be the most helpful way to describe this behavior.  (Sphinx can also auto-generate nice HTML renderings for tables embedded in docstrings.)

File: zipline/_protocol.pyx [hunk 38]
```diff
@@ -76,6 +81,289 @@ cdef class BarData:
...
+        If a list of assets and a single field are passed in, a pandas Series
+        is returned whose indices are the assets, and whose values are scalar
+        values for each asset for the given field.
+
+        If a list of assets and a list of fields are passed in, a pandas
+        DataFrame is returned, indexed by asset.  The columns are the requested
+        fields, filled with the scalar values for each asset for each field.
+
+        "price" returns the last known close price of the asset.  If there is
+        no last known value, NaN is returned.  If a value is found, and we
```
Comment (ssanderson) [inline zipline/_protocol.pyx]: Might be worth clarifying that `NaN` is only returned if we've never seen a trade.

File: zipline/_protocol.pyx [hunk 39]
```diff
@@ -76,6 +81,289 @@ cdef class BarData:
...
+        had to cross an adjustment boundary (split, dividend, etc) to get it,
```
Comment (ssanderson) [inline zipline/_protocol.pyx]: :+1: for clarifying the edge case here.

File: zipline/_protocol.pyx [hunk 40]
```diff
@@ -76,6 +81,289 @@ cdef class BarData:
...
+        the value is adjusted before being returned.
+
+        "last_traded" returns the date of the last trade event of the asset,
+        even if the asset has stopped trading. If there is no last known value,
+        NaT is returned.
```
Comment (ssanderson) [inline zipline/_protocol.pyx]: Is this numpy.NaT or pd.NaT?  (They actually have pretty different behavior.)

File: zipline/_protocol.pyx [hunk 41]
```diff
@@ -76,6 +81,289 @@ cdef class BarData:
             self.data_frequency
         )
 
+    def spot_value(self, assets, fields):
+        """
+        Returns the spot value of the given assets for the given fields
+        at the current simulation time.  Spot values are the as-traded price
```
Comment (ssanderson) [inline zipline/_protocol.pyx]: Based on the comment below on `price`, it seems like this isn't quite true anymore?

File: zipline/_protocol.pyx [hunk 42]
```diff
@@ -76,6 +81,289 @@ cdef class BarData:
...
+        Parameters
+        ----------
+        assets : Asset or iterable of Assets
+
+        fields : string or iterable of strings.  Valid values are: "price",
+            "last_traded", "open", "high", "low", "close", "volume"
+
+        Returns
+        -------
+        Float, pandas Series, or pandas DataFrame.  See notes
```
Comment (ssanderson) [inline zipline/_protocol.pyx]: This is an Int if we request a scalar volume, right?

File: zipline/data/us_equity_pricing.py (hunk 24 above)
Comment (jbredeche) [inline zipline/data/us_equity_pricing.py]: I rewrote that chunk of code, agreed with @ssanderson that the try/except was too broad.  Now, we simply go back to the last trading day before the asset end date, if needed.

```
        if day >= asset.end_date:
            # go back to one day before the asset ended
            search_day = self._calendar[
                self._calendar.get_loc(asset.end_date) - 1
            ]
        else:
            search_day = day
```

File: zipline/sources/benchmark_source.py (hunk 26 above)
Comment (jbredeche) [inline zipline/sources/benchmark_source.py]: we were passing an `int` into `data_portal.get_history_window`, and I wanted to avoid that.  Also cleaned up a bit of the initialization logic.

File: zipline/utils/test_utils.py (hunk 28 above)
Comment (jbredeche) [inline zipline/utils/test_utils.py]: I think my version is more readable, and it's fast enough

File: zipline/_protocol.pyx (hunk 30 above)
Comment (jbredeche) [inline zipline/_protocol.pyx]: fixed

File: zipline/finance/blotter.py (hunk 32 above)
Comment (jbredeche) [inline zipline/finance/blotter.py]: good catch, fixed

File: zipline/finance/slippage.py (hunk 33 above)
Comment (jbredeche) [inline zipline/finance/slippage.py]: I don't see that here.

File: zipline/finance/slippage.py (hunk 34 above)
Comment (jbredeche) [inline zipline/finance/slippage.py]: we could, but I think it adds a lot of power for the user to have access to `bar_data` when building a custom slippage function.  They can do `history` calls, check volatility, etc.

File: tests/test_bar_data.py (hunk 1 above)
Comment (jbredeche) [inline tests/test_bar_data.py]: thanks, done

File: zipline/_protocol.pyx (hunk 7 above)
Comment (jbredeche) [inline zipline/_protocol.pyx]: good idea, done

File: zipline/utils/test_utils.py (hunk 36 above)
Comment (jbredeche) [inline zipline/utils/test_utils.py]: agreed, fixed

File: zipline/_protocol.pyx (hunk 38 above)
Comment (jbredeche) [inline zipline/_protocol.pyx]: I'm clarifying, but `NaN` is returned if we've never seen a trade, OR if the asset is dead.

File: zipline/_protocol.pyx (hunk 40 above)
Comment (jbredeche) [inline zipline/_protocol.pyx]: `pd.NaT`

File: zipline/utils/test_utils.py (hunk 36 above)
Comment (jbredeche) [inline zipline/utils/test_utils.py]: that change triggered some warnings:
```
A value is trying to be set on a copy of a slice from a DataFrame

See the the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._setitem_with_indexer(indexer, value)

```

leaving it alone for now.

File: zipline/_protocol.pyx (hunk 42 above)
Comment (jbredeche) [inline zipline/_protocol.pyx]: yes, will clarify

File: zipline/_protocol.pyx (hunk 41 above)
Comment (jbredeche) [inline zipline/_protocol.pyx]: yeah.  but the vast majority of the time, spot values aren't adjusted, and I want to reinforce that to users.  I'll fiddle with the wording.
